### PR TITLE
WIP: Update documentation and the test scripts for the --dry-run and --server-dry-run

### LINKED ISSUE #86526: document server-side dry run is GA
**What would you like to be added**:

Server-side dry-run is being promoted from beta to GA in v1.18.

For the server-side dry-run integration in kubectl, we are making a few changes:

- we're extending the `--dry-run` flag to support the options `client`, `server` and `none`
- we're deprecating no value for the `--dry-run` flag, such that setting a value will be required in the future
- we're deprecating the boolean values for the `--dry-run` flag and will remove these values in the future
- we're deprecating the `kubectl apply` `--server-dry-run` flag

We need to document these changes accordingly on the website and in comments and tests.

Documentation includes:

- [x] document server-side dry-run feature state as stable for v1.18 (update Hugo shortcode) https://github.com/kubernetes/website/pull/19549
- [x] dry-run auth requirements @julianvmodesto https://github.com/kubernetes/website/pull/18235
- [ ] document all API requests, including server-side dry-run, are single-resource requests https://github.com/kubernetes/kubernetes/issues/83562 https://github.com/kubernetes/kubernetes/issues/86026 https://github.com/kubernetes/website/pull/19550
- [x] deprecation notices for `--dry-run`: covered by release notes
- [x] deprecation notice for `--server-dry-run`: covered by release notes
- [x] all usage and examples on the website https://github.com/kubernetes/website/pull/19549
- [x] all usage in code comments and tests

**Why is this needed**:

Server-side dry-run is graduating to GA, so we need to document usage and examples for users for how to use the feature.

Because we're extending an existing flag and deprecating aspects of this existing feature, we need to have consistent documentation for the flag that reflects the changes.

https://github.com/kubernetes/kubernetes/issues/85652
GA umbrella issue: https://github.com/kubernetes/kubernetes/issues/68514
KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/0015-dry-run.md

cc @apelisse 
/sig cli
/sig api-machinery
/priority important-soon
/help

### PR DESCRIPTION
**Pull request by bharaththiruveedula-zz**
**What type of PR is this?**
> /kind documentation

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #86526

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: cmd/kubeadm/app/cmd/token_test.go [hunk 1]
```diff
@@ -226,13 +226,13 @@ func TestNewCmdToken(t *testing.T) {
 		},
 		{
 			name:          "valid: delete from --kubeconfig",
-			args:          []string{"delete", "abcdef.1234567890123456", "--dry-run", "--kubeconfig=" + fullPath},
+			args:          []string{"delete", "abcdef.1234567890123456", "--dry-run=client", "--kubeconfig=" + fullPath},
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/token_test.go]: hi, i don't think the cmd/kubeadm changes in this PR are valid.
did you happen to use an automated search and replace query to replace these?

please note that different components (kubelet, kubeadm, or even their sub-commands) might have different dry-run implementations.

/hold

File: cmd/kubelet/app/options/options.go [hunk 2]
```diff
@@ -55,6 +55,7 @@ const defaultRootDir = "/var/lib/kubelet"
 type KubeletFlags struct {
 	KubeConfig          string
 	BootstrapKubeconfig string
+	VersionJSON         bool
```
Comment (apelisse) [inline cmd/kubelet/app/options/options.go]: These and the following related change, probably shouldn't be included in that PR.

File: staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go [hunk 3]
```diff
@@ -386,6 +386,16 @@ func (o *ApplyOptions) Run() error {
 	if len(infos) == 0 {
 		return fmt.Errorf("no objects passed to apply")
 	}
+
```
Comment (apelisse) [inline staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go]: You should revert the changes in this file. As we're introducing the new flag we're also adding these warning where they should be. thanks.

File: cmd/kubeadm/app/cmd/token_test.go (hunk 1 above)
Comment (apelisse) [reply cmd/kubeadm/app/cmd/token_test.go]: Yes, please remove this change since I don't think it applies to the right command.

File: staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go [hunk 4]
```diff
@@ -386,6 +386,15 @@ func (o *ApplyOptions) Run() error {
 	if len(infos) == 0 {
 		return fmt.Errorf("no objects passed to apply")
 	}
+	// Add Deprecation notices for both --dry-run and --server-dry-run
```
Comment (apelisse) [inline staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go]: Remove all of these changes in this file.

Comment (apelisse) [reply staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go]: There is another PR that will introduce these deprecations properly.

Comment (apelisse) [reply staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go]: #87580

**Review by apelisse — COMMENTED** [review]
Almost looks good, thanks!

**Comment by apelisse** [comment]
This is very close to merge, I would really appreciate the few comments to be fixed! Thanks

**Comment by julianvmodesto** [comment]
@bharaththiruveedula could you please rebase master into your change?

**Comment by bharaththiruveedula** [comment]
@julianvmodesto Sure I will do that. Sorry for late reply

**Comment by apelisse** [comment]
@julianvmodesto I think you've taken this and it got merged right? Please re-open if not, thanks.
/close
